util/locallylock/lock.go

```diff
@@ -0,0 +1,52 @@
+package locallylock
+
+import (
+	"context"
+	"sync"
+)
+
+type Lock struct {
+	m       *sync.Mutex
+	c       *sync.Cond
+	current string
+	ctx     context.Context
+}
+
+func New() *Lock {
+	m := sync.Mutex{}
+	c := sync.NewCond(&m)
+
+	return &Lock{
+		m: &m,
+		c: c,
+	}
+}
+
+func (l *Lock) Lock(ctx context.Context, name string) {
+	l.m.Lock()
+	for {
+		if l.ctx != nil {
+			select {
+			case <-l.ctx.Done():
+				// unlock locks from closed connections
+				l.current = ""
```

Comment:
Not sure if this is absolutely needed - but feel like there might be an edge case that we could be missing?

```suggestion
				l.current = ""
				l.ctx = nil
```